File: server/api/schemas/order.py
from pydantic import BaseModel
from pydantic.utils import GetterDict
from typing import List

from .project import Project
from .user import User
from .distributor import Distributor
from .contactPerson import ContactPerson

class OrderItemGetter(GetterDict):
    def get(self, key: str, default = None):
        if key in {'id', 'name', 'model', 'category', 'quantity', 'comment', 'project', 'currency'}:
            return getattr(self._obj.item, key)
        else:
            return super(OrderItemGetter, self).get(key, default)

class OrderItem(BaseModel):
    id: int
    name: str
    model: str
    category: str
    quantity: int
    comment: str
    project: Project
    price: float
    currency: str

    class Config:
        orm_mode = True
        getter_dict = OrderItemGetter

class OrderItemCreate(BaseModel):
    Item_id: int
    order_id: int
    quantity: int
    price: float
    currency: str

class OrderItemEdit(BaseModel):
    id: int
    quantity: int
    price: float

class Order(BaseModel):
    id: int
    user: User
    distributor: Distributor
    contactPerson: ContactPerson
    dateAndTime: str
    archived: bool
    items: List[OrderItem]

    class Config:
        orm_mode = True

class OrderCreate(BaseModel):
    idUser: int
    idDistributor: int
    idContactPerson: int
    archived: bool

class OrderEdit(BaseModel):
    id: int
    idUser: int
    idDistributor: int
    idContactPerson: int
    archived: bool

File: client/web/src/app/shared/models/order.ts
import { IContactPerson, IDistributor, IUser } from ".";

export interface IPOrder {
  items?: IOrder[];
  previous_page?: number;
  next_page?: number;
  has_previous?: boolean;
  has_next?: boolean;
  total?: number;
  pages?: number;
}

export interface IOrder {
  id?: number;
  idUser?: number;
  user?: IUser;
  idDistributor?: number;
  distributor?: IDistributor;
  idContactPerson?: number;
  contactPerson?: IContactPerson;
  dateAndTime?: string;
  archived?: boolean;
  items?: IOrderItem[];
}

export interface IOrderItem {
  Item_id?: number;
  order_id?: number;
  quantity?: number;
  price?: number;
  currency?: string;
}
